pkg/datastore/headers.go

```diff
@@ -0,0 +1,433 @@
+		return json.Unmarshal(val, &h)
+	})
+
+	return h, err
+}
+
+func (s *Datastore) PutHeader(ctx context.Context, hr structs.HR, ttl time.Duration) (err error) {
+	if err := storeHeader(s.Badger, hr, ttl); err != nil {
+		return err
+	}
+
+	newlyCreated, err := s.hc.Add(uint64(hr.Slot), hr.HeaderAndTrace)
+	if err != nil {
+		return err
+	}
+
+	if !newlyCreated {
+		return // success
+	}
+
+	// check and load keys if exists
+	return s.loadKeysAndCleanup(ctx, uint64(hr.Slot))
+}
+
+func (s *Datastore) loadKeysAndCleanup(ctx context.Context, slot uint64) error {
+	s.l.Lock()
+	defer s.l.Unlock()
+
+	// need to load keys to memory
+	data, err := s.TTLStorage.Get(ctx, HeaderKey(slot))
+	if err != nil {
+		if errors.Is(err, ds.ErrNotFound) {
+			return nil // success - key is empty, nothing to load
+		}
+		return err
+	}
+	var h []structs.HeaderAndTrace
+	if err := json.Unmarshal(data, &h); err != nil {
+		return err
+	}
+
+	return s.hc.AddMultiple(slot, h)
+}
+
+func storeHeader(s Badger, h structs.HR, ttl time.Duration) error {
+	txn := s.NewTransaction(true)
+	defer txn.Discard()
+
+	// we don't need to lock here, as the value would be always different from different block
+	if err := txn.SetEntry(badger.NewEntry(HeaderKeyContent(h.Slot, h.Trace.BlockHash.String()).Bytes(), h.Marshaled).WithTTL(ttl)); err != nil {
+		return err
+	}
+
+	slot := make([]byte, 8)
+	binary.LittleEndian.PutUint64(slot, uint64(h.Slot))
+
+	if err := txn.SetEntry(badger.NewEntry(HeaderHashKey(h.Header.BlockHash).Bytes(), slot).WithTTL(ttl)); err != nil {
+		return err
+	}
+
+	// not needed every time
+	if err := txn.SetEntry(badger.NewEntry(HeaderNumKey(h.Header.BlockNumber).Bytes(), slot).WithTTL(ttl)); err != nil {
+		return err
+	}
+
+	return txn.Commit()
+}
+
+func (s *Datastore) GetHeadersBySlot(ctx context.Context, slot uint64) ([]structs.HeaderAndTrace, error) {
+	el, _ := s.hc.GetContent(slot, slot, 1)
+	if el != nil {
+		return el, nil
+	}
+
+	data, err := s.TTLStorage.Get(ctx, HeaderKey(slot))
+	if err != nil && errors.Is(err, ds.ErrNotFound) {
+		return el, err
+	}
+
+	el = []structs.HeaderAndTrace{}
+	if err = json.Unmarshal(data, &el); err != nil {
+		return el, err
+	}
+
+	return el, err
+}
+
+func (s *Datastore) GetHeadersByBlockNum(ctx context.Context, blockNumber uint64) ([]structs.HeaderAndTrace, error) {
+	slot, err := s.TTLStorage.Get(ctx, HeaderNumKey(blockNumber))
+	if err != nil {
+		return nil, err
+	}
+	return s.GetHeadersBySlot(ctx, binary.LittleEndian.Uint64(slot))
+}
+
+func (s *Datastore) GetHeadersByBlockHash(ctx context.Context, hash types.Hash) ([]structs.HeaderAndTrace, error) {
+	slot, err := s.TTLStorage.Get(ctx, HeaderHashKey(hash))
+	if err != nil {
+		return nil, err
+	}
+	return s.GetHeadersBySlot(ctx, binary.LittleEndian.Uint64(slot))
+}
+
+func (s *Datastore) GetLatestHeaders(ctx context.Context, limit uint64) ([]structs.HeaderAndTrace, error) {
+	ls := s.hc.GetLatestSlot()
+	stop := ls - uint64(24*time.Hour/time.Second*12)
```

Comment:
Here we should add a check for underflow

Reply:
hm? It's exactly the same way how it was no ? :) You're thinking of math.Floor()?

Reply:
Okay, don't worry, we can fix this in another PR. There is a PR opened that fixes this. But not for this PR.